debezium/loadgen/generate_load.py

```diff
@@ -441,104 +446,128 @@ async def run_command_server():
+                "database.server.name": mysql_host,
+                "database.server.id": '1234',
+                "database.history.kafka.bootstrap.servers": kafka_endpoint,
+                "database.history.kafka.topic": "mysql-history",
+                "time.precision.mode": "connect"
+            }
         }
-    }
-)
-
-# Initialize Kafka
-
-producers = []
-db_resources = []
-exit_code = 0
-
-try:
-    KafkaAdminClient(bootstrap_servers=kafka_endpoint).create_topics([
-        NewTopic(
-            name="pageviews",
-            num_partitions=kafka_partitions,
-            replication_factor=1
-        )
-    ])
-
-    for pi in range(max_parallel_purchases + max_parallel_pageviews):
-        producer = KafkaProducer(
-            bootstrap_servers = [kafka_endpoint],
-            acks = kafka_producer_acks,
-            batch_size = kafka_batch_size,
-            max_in_flight_requests_per_connection = max_parallel_purchases + max_parallel_pageviews,
-            linger_ms = kafka_linger_ms,
-            value_serializer = lambda x: json.dumps(x).encode('utf-8')
-        )
-        producers.append(producer)
+    )
 
-    for di in range(max_parallel_purchases):
-        connection = connect(host=mysql_host, user=mysql_user, password=mysql_pass)
-        cursor = connection.cursor()
-        db_resources.append((connection, cursor))
-
-    (connection, cursor) = db_resources[0]
-    initialize_database(connection, cursor)
-    seed_data(connection, cursor)
-    item_prices = get_item_prices(cursor)
-    async def shutdown():
-        for task in asyncio.Task.all_tasks():
-            task.cancel()
-    async def run_async():
-        event_loop = asyncio.get_event_loop()
-        signals = (signal.SIGHUP, signal.SIGTERM, signal.SIGINT)
-        async def handle_signal(signal):
-            await shutdown()
-        for s in signals:
-            event_loop.add_signal_handler(
-                s, lambda s=s: asyncio.create_task(handle_signal(s)))
-        with concurrent.futures.ThreadPoolExecutor(
-                max_workers = max_parallel_purchases + max_parallel_pageviews
-        ) as executor:
-            await asyncio.gather(
-                loop_purchases(executor, db_resources, producers, item_prices),
-                loop_random_pageviews(executor, producers),
-                run_command_server()
-            )
-    log("Starting simulation loops.")
+    exit_code = 0
     try:
-        asyncio.run(run_async())
+        # Setup our topic; necessary for num_partitions > 1.
+        KafkaAdminClient(bootstrap_servers=kafka_endpoint).create_topics([
+            NewTopic(
+                name="pageviews",
+                num_partitions=kafka_partitions,
+                replication_factor=1
+            )
+        ])
+
+        # With python 3.10 we will be able to use parenthesis around
+        # 'with' with multiple 'as', and get rid
+        # of the backslashes.
+        with \
```

Comment:
Do you need to have this gated on Kafka only?

Reply:
This is database initialization code in main; is independent of what the executors (which are separate processes) do.
The initialization code that gates on kafka is run on the forked sub-processes of the process pool executor.
I've changed the names of the methods and global variables that are executor specific to all start with `executor_` in the hope that that makes it more clear.